File: sdk/python/microsandbox/base_sandbox.py
"""
Base sandbox implementation for the Microsandbox Python SDK.
"""

import asyncio
import os
import uuid
from abc import ABC, abstractmethod
from contextlib import asynccontextmanager
from typing import Dict, List, Optional

import aiohttp
from dotenv import load_dotenv

from .command import Command
from .metrics import Metrics


class BaseSandbox(ABC):
    """
    Base sandbox environment for executing code safely.

    This class provides the base interface for interacting with the Microsandbox server.
    It handles common functionality like sandbox creation, management, and communication.
    """

    def __init__(
        self,
        server_url: str = None,
        name: Optional[str] = None,
        api_key: Optional[str] = None,
    ):
        """
        Initialize a base sandbox instance.

        Args:
            server_url: URL of the Microsandbox server. If not provided, will check MSB_SERVER_URL environment variable, then fall back to default.
            name: Optional name for the sandbox. If not provided, a random name will be generated.
            api_key: API key for Microsandbox server authentication. If not provided, it will be read from MSB_API_KEY environment variable.
        """
        # Only try to load .env if MSB_API_KEY is not already set
        if "MSB_API_KEY" not in os.environ:
            # Ignore errors if .env file doesn't exist
            try:
                load_dotenv()
            except Exception:
                pass

        self._server_url = server_url or os.environ.get(
            "MSB_SERVER_URL", "http://127.0.0.1:5555"
        )
        self._name = name or f"sandbox-{uuid.uuid4().hex[:8]}"
        self._api_key = api_key or os.environ.get("MSB_API_KEY")
        self._session = None
        self._is_started = False

    @abstractmethod
    async def get_default_image(self) -> str:
        """
        Get the default Docker image for this sandbox type.

        Returns:
            A string containing the Docker image name and tag
        """
        pass

    @classmethod
    @asynccontextmanager
    async def create(
        cls,
        server_url: str = None,
        name: Optional[str] = None,
        api_key: Optional[str] = None,
    ):
        """
        Create and initialize a new sandbox as an async context manager.

        Args:
            server_url: URL of the Microsandbox server. If not provided, will check MSB_SERVER_URL environment variable, then fall back to default.
            name: Optional name for the sandbox. If not provided, a random name will be generated.
            api_key: API key for Microsandbox server authentication. If not provided, it will be read from MSB_API_KEY environment variable.

        Returns:
            An instance of the sandbox ready for use
        """
        # Only try to load .env if MSB_API_KEY is not already set
        if "MSB_API_KEY" not in os.environ:
            # Ignore errors if .env file doesn't exist
            try:
                load_dotenv()
            except Exception:
                pass

        sandbox = cls(
            server_url=server_url,
            name=name,
            api_key=api_key,
        )
        try:
            # Create HTTP session
            sandbox._session = aiohttp.ClientSession()
            # Start the sandbox
            await sandbox.start()
            yield sandbox
        finally:
            # Stop the sandbox
            await sandbox.stop()
            # Close the HTTP session
            if sandbox._session:
                await sandbox._session.close()
                sandbox._session = None

    async def start(
        self,
        image: Optional[str] = None,
        memory: int = 512,
        cpus: float = 1.0,
        volumes: Optional[List[str]] = None,
        ports: Optional[List[str]] = None,
        envs: Optional[List[str]] = None,
        depends_on: Optional[List[str]] = None,
        workdir: Optional[str] = None,
        shell: Optional[str] = None,
        scripts: Optional[Dict[str, str]] = None,
        exec: Optional[str] = None,
        timeout: float = 180.0,
    ) -> None:
        """
        Start the sandbox container.

        Args:
            image: Docker image to use for the sandbox (defaults to language-specific image)
            memory: Memory limit in MB
            cpus: CPU limit (will be rounded to nearest integer)
            volumes: Volumes to mount
            ports: Ports to expose
            envs: Environment variables to use
            depends_on: Sandboxes to depend on
            workdir: Working directory to use
            shell: Shell to use
            scripts: Scripts that can be run
            exec: Exec command to run
            timeout: Maximum time in seconds to wait for the sandbox to start (default: 180 seconds)

        Raises:
            RuntimeError: If the sandbox fails to start
            TimeoutError: If the sandbox doesn't start within the specified timeout
        """
        if self._is_started:
            return

        sandbox_image = image or await self.get_default_image()
        config = {
            "image": sandbox_image,
            "memory": memory,
            "cpus": int(round(cpus)),
        }
        if volumes is not None:
            config["volumes"] = volumes
        if ports is not None:
            config["ports"] = ports
        if envs is not None:
            config["envs"] = envs
        if depends_on is not None:
            config["depends_on"] = depends_on
        if workdir is not None:
            config["workdir"] = workdir
        if shell is not None:
            config["shell"] = shell
        if scripts is not None:
            config["scripts"] = scripts
        if exec is not None:
            config["exec"] = exec

        request_data = {
            "jsonrpc": "2.0",
            "method": "sandbox.start",
            "params": {
                "sandbox": self._name,
                "config": config,
            },
        }

        headers = {"Content-Type": "application/json"}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"

        try:
            # Set a client-side timeout that's a bit longer than the server-side timeout
            # to account for network latency and processing time
            client_timeout = aiohttp.ClientTimeout(total=timeout + 30)

            async with self._session.post(
                f"{self._server_url}/api/v1/sandbox/start",
                json=request_data,
                headers=headers,
                timeout=client_timeout,
            ) as response:
                if response.status != 200:
                    error_text = await response.text()
                    raise RuntimeError(f"Failed to start sandbox: {error_text}")

                response_data = await response.json()

                # Check the message - it might indicate the sandbox is still initializing
                message = response_data.get("message", "")
                if isinstance(message, str) and "timed out waiting" in message:
                    # Server timed out but still started the sandbox
                    # We'll raise a warning but still consider it started
                    import warnings

                    warnings.warn(f"Sandbox start warning: {message}")

                self._is_started = True
        except aiohttp.ClientError as e:
            if isinstance(e, asyncio.TimeoutError):
                raise TimeoutError(
                    f"Timed out waiting for sandbox to start after {timeout} seconds"
                ) from e
            raise RuntimeError(f"Failed to communicate with Microsandbox server: {e}")

    async def stop(self) -> None:
        """
        Stop the sandbox container.

        Raises:
            RuntimeError: If the sandbox fails to stop
        """
        if not self._is_started:
            return

        request_data = {
            "jsonrpc": "2.0",
            "method": "sandbox.stop",
            "params": {"sandbox": self._name},
            "id": str(uuid.uuid4()),
        }

        headers = {"Content-Type": "application/json"}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"

        try:
            async with self._session.post(
                f"{self._server_url}/api/v1/sandbox/stop",
                json=request_data,
                headers=headers,
            ) as response:
                if response.status != 200:
                    error_text = await response.text()
                    raise RuntimeError(f"Failed to stop sandbox: {error_text}")

                self._is_started = False
        except aiohttp.ClientError as e:
            raise RuntimeError(f"Failed to communicate with Microsandbox server: {e}")

    @abstractmethod
    async def run(self, code: str):
        """
        Execute code in the sandbox.

        Args:
            code: Code to execute

        Returns:
            An Execution object representing the executed code

        Raises:
            RuntimeError: If execution fails
        """
        pass

    @property
    def command(self):
        """
        Access the command namespace for executing shell commands in the sandbox.

        Returns:
            A Command instance bound to this sandbox
        """
        return Command(self)

    @property
    def metrics(self):
        """
        Access the metrics namespace for retrieving sandbox metrics.

        Returns:
            A Metrics instance bound to this sandbox
        """
        return Metrics(self)

File: sdk/javascript/src/base-sandbox.ts
/**
 * Base sandbox implementation for the Microsandbox TypeScript SDK.
 */

import { v4 as uuidv4 } from "uuid";
import fetch from "node-fetch";
import * as dotenv from "dotenv";

import { Command } from "./command";
import { Metrics } from "./metrics";
import { Execution } from "./execution";
import { SandboxOptions } from "./types";

export abstract class BaseSandbox {
  protected _serverUrl: string;
  protected _name: string;
  protected _apiKey: string | undefined;
  protected _isStarted: boolean = false;

  /**
   * Initialize a base sandbox instance.
   *
   * @param options - Configuration options for the sandbox
   */
  constructor(options?: SandboxOptions) {
    // Try to load from .env if MSB_API_KEY is not already set
    if (!process.env.MSB_API_KEY) {
      try {
        dotenv.config();
      } catch (error) {
        // Ignore errors if .env file doesn't exist
      }
    }

    this._serverUrl =
      options?.serverUrl ||
      process.env.MSB_SERVER_URL ||
      "http://127.0.0.1:5555";
    this._name = options?.name || `sandbox-${uuidv4().substring(0, 8)}`;
    this._apiKey = options?.apiKey || process.env.MSB_API_KEY;
  }

  /**
   * Create and initialize a new sandbox instance.
   *
   * This is the base implementation that subclasses should call.
   *
   * @param ctor - The constructor function for the subclass
   * @param options - Configuration options for the sandbox
   * @returns A Promise resolving to a new sandbox instance
   */
  protected static async createBase<T extends BaseSandbox>(
    ctor: new (options?: SandboxOptions) => T,
    options?: SandboxOptions,
  ): Promise<T> {
    // Try to load from .env if MSB_API_KEY is not already set
    if (!process.env.MSB_API_KEY) {
      try {
        dotenv.config();
      } catch (error) {
        // Ignore errors if .env file doesn't exist
      }
    }

    const sandbox = new ctor(options);

    // Start the sandbox
    await sandbox.start(
      options?.image,
      options?.memory,
      options?.cpus,
      options?.timeout,
    );

    return sandbox;
  }

  // Abstract static method signature that subclasses must implement
  static create(options?: SandboxOptions): Promise<BaseSandbox> {
    throw new Error("Static method 'create' must be implemented by subclass");
  }

  /**
   * Get the default Docker image for this sandbox type.
   *
   * @returns A string containing the Docker image name and tag
   */
  abstract getDefaultImage(): Promise<string>;

  /**
   * Start the sandbox container.
   *
   * @param image - Docker image to use for the sandbox (defaults to language-specific image)
   * @param memory - Memory limit in MB
   * @param cpus - CPU limit (will be rounded to nearest integer)
   * @param timeout - Maximum time in seconds to wait for the sandbox to start (default: 180 seconds)
   *
   * @throws Error if the sandbox fails to start
   * @throws Error if the sandbox doesn't start within the specified timeout
   */
  async start(
    image?: string,
    memory: number = 512,
    cpus: number = 1.0,
    timeout: number = 180.0,
  ): Promise<void> {
    if (this._isStarted) {
      return;
    }

    const sandboxImage = image || (await this.getDefaultImage());
    const requestData = {
      jsonrpc: "2.0",
      method: "sandbox.start",
      params: {
        sandbox: this._name,
        config: {
          image: sandboxImage,
          memory,
          cpus: Math.round(cpus),
        },
      },
    };

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };

    if (this._apiKey) {
      headers["Authorization"] = `Bearer ${this._apiKey}`;
    }

    try {
      const response = await fetch(`${this._serverUrl}/api/v1/sandbox/start`, {
        method: "POST",
        headers,
        body: JSON.stringify(requestData),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Failed to start sandbox: ${errorText}`);
      }

      const responseData = await response.json();

      if ("error" in responseData) {
        throw new Error(
          `Failed to start sandbox: ${responseData.error.message}`,
        );
      }

      // Check the result message - it might indicate the sandbox is still initializing
      const result = responseData.result;
      if (typeof result === "string" && result.includes("timed out waiting")) {
        // Server timed out but still started the sandbox
        // We'll log a warning but still consider it started
        console.warn(`Sandbox start warning: ${result}`);
      }

      this._isStarted = true;
    } catch (e) {
      if (e instanceof Error) {
        if (e.message.includes("timeout")) {
          throw new Error(
            `Timed out waiting for sandbox to start after ${timeout} seconds`,
          );
        }
        throw new Error(
          `Failed to communicate with Microsandbox server: ${e.message}`,
        );
      }
      throw new Error(
        "Failed to communicate with Microsandbox server: Unknown error",
      );
    }
  }

  /**
   * Stop the sandbox container.
   *
   * @throws Error if the sandbox fails to stop
   */
  async stop(): Promise<void> {
    if (!this._isStarted) {
      return;
    }

    const requestData = {
      jsonrpc: "2.0",
      method: "sandbox.stop",
      params: {
        sandbox: this._name,
      },
      id: uuidv4(),
    };

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };

    if (this._apiKey) {
      headers["Authorization"] = `Bearer ${this._apiKey}`;
    }

    try {
      const response = await fetch(`${this._serverUrl}/api/v1/sandbox/stop`, {
        method: "POST",
        headers,
        body: JSON.stringify(requestData),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Failed to stop sandbox: ${errorText}`);
      }

      const responseData = await response.json();

      if ("error" in responseData) {
        throw new Error(
          `Failed to stop sandbox: ${responseData.error.message}`,
        );
      }

      this._isStarted = false;
    } catch (e) {
      if (e instanceof Error) {
        throw new Error(
          `Failed to communicate with Microsandbox server: ${e.message}`,
        );
      }
      throw new Error(
        "Failed to communicate with Microsandbox server: Unknown error",
      );
    }
  }

  /**
   * Execute code in the sandbox.
   *
   * @param code - Code to execute
   * @returns An Execution object representing the executed code
   * @throws Error if execution fails
   */
  abstract run(code: string): Promise<Execution>;

  /**
   * Access the command namespace for executing shell commands in the sandbox.
   */
  get command(): Command {
    return new Command(this);
  }

  /**
   * Access the metrics namespace for retrieving sandbox metrics.
   */
  get metrics(): Metrics {
    return new Metrics(this);
  }

  /**
   * Check if the sandbox is started.
   */
  get isStarted(): boolean {
    return this._isStarted;
  }

  /**
   * Get the server URL.
   */
  get serverUrl(): string {
    return this._serverUrl;
  }

  /**
   * Get the sandbox name.
   */
  get name(): string {
    return this._name;
  }

  /**
   * Get the API key.
   */
  get apiKey(): string | undefined {
    return this._apiKey;
  }
}
